authority/provisioner/k8sSA.go

```diff
@@ -0,0 +1,245 @@
+		// creating the authn until token validation time?
+		if err := checkAccess(k8s.AuthorizationV1()); err != nil {
+			return errors.Wrapf(err, "error verifying access to kubernetes authz service for provisioner %s", p.GetID())
+		}
+
+		p.kauthn = k8s.AuthenticationV1()
+	*/
+
+	// Update claims with global ones
+	if p.claimer, err = NewClaimer(p.Claims, config.Claims); err != nil {
+		return err
+	}
+
+	p.audiences = config.Audiences
+	return err
+}
+
+// authorizeToken performs common jwt authorization actions and returns the
+// claims for case specific downstream parsing.
+// e.g. a Sign request will auth/validate different fields than a Revoke request.
+func (p *K8sSA) authorizeToken(token string, audiences []string) (*k8sSAPayload, error) {
+	jwt, err := jose.ParseSigned(token)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error parsing token")
+	}
+
+	var (
+		valid  bool
+		claims k8sSAPayload
+	)
+	if p.pubKeys == nil {
+		return nil, errors.New("TokenReview API integration not implemented")
+		/*
+			tr := kauthnApi.TokenReview{Spec: kauthnApi.TokenReviewSpec{Token: string(token)}}
+			rvw, err := p.kauthn.TokenReviews().Create(&tr)
+			if err != nil {
+				return nil, errors.Wrap(err, "error using kubernetes TokenReview API")
+			}
+			if rvw.Status.Error != "" {
+				return nil, errors.Errorf("error from kubernetes TokenReviewAPI: %s", rvw.Status.Error)
+			}
+			if !rvw.Status.Authenticated {
+				return nil, errors.New("error from kubernetes TokenReviewAPI: token could not be authenticated")
+			}
+			if err = jwt.UnsafeClaimsWithoutVerification(&claims); err != nil {
+				return nil, errors.Wrap(err, "error parsing claims")
+			}
+		*/
+	}
+	for _, pk := range p.pubKeys {
+		if err = jwt.Claims(pk, &claims); err == nil {
+			valid = true
+			break
+		}
+	}
+	if !valid {
+		return nil, errors.New("error validating token and extracting claims")
+	}
+
+	// According to "rfc7519 JSON Web Token" acceptable skew should be no
+	// more than a few minutes.
+	if err = claims.Validate(jose.Expected{
+		Issuer: k8sSAIssuer,
+	}); err != nil {
+		return nil, errors.Wrapf(err, "invalid token claims")
+	}
+
+	if claims.Subject == "" {
+		return nil, errors.New("token subject cannot be empty")
+	}
+
+	return &claims, nil
+}
+
+// AuthorizeRevoke returns an error if the provisioner does not have rights to
+// revoke the certificate with serial number in the `sub` property.
+func (p *K8sSA) AuthorizeRevoke(token string) error {
+	_, err := p.authorizeToken(token, p.audiences.Revoke)
+	return err
+}
+
+// AuthorizeSign validates the given token.
+func (p *K8sSA) AuthorizeSign(ctx context.Context, token string) ([]SignOption, error) {
+	_, err := p.authorizeToken(token, p.audiences.Sign)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check for SSH sign-ing request.
+	if MethodFromContext(ctx) == SignSSHMethod {
+		return nil, errors.New("ssh certificates not enabled for k8s ServiceAccount provisioners")
+	}
```

Comment:
I suppose there's no point in using SSH because you should use `kubectl exec`, right?

Reply:
That was my thinking, but might be something to have @mmalone weigh in on as well.

Reply:
I think this is fine for now. I can't think of a k8s/SSH use case. If someone comes along with one we can reconsider.

Reply:
Sounds good.